File: rust/src/engine.rs
//! Motor de inferencia en Rust puro. Reproduce exactamente `src/portable/engine.py`.
//! Carga pesos .safetensors (entrenados con MLX) y corre en cualquier CPU.

use std::collections::HashMap;
use std::path::Path;

use safetensors::SafeTensors;

/// Config del modelo (viene del sidecar .json que guarda el entrenamiento).
pub struct Config {
    pub vocab_size: usize,
    pub block_size: usize,
    pub dim: usize,
    pub n_layers: usize,
    pub n_heads: usize,
}

/// Pesos: nombre -> vector plano f32 (row-major, igual que safetensors/PyTorch).
pub struct Model {
    w: HashMap<String, Vec<f32>>,
    pub cfg: Config,
    head_dim: usize,
}

fn erf(x: f32) -> f32 {
    // Abramowitz & Stegun 7.1.26 (mismo que engine.py, error < 1.5e-7).
    let sign = if x < 0.0 { -1.0 } else { 1.0 };
    let x = x.abs();
    let t = 1.0 / (1.0 + 0.3275911 * x);
    let y = 1.0
        - (((((1.061405429 * t - 1.453152027) * t) + 1.421413741) * t - 0.284496736) * t
            + 0.254829592)
            * t
            * (-x * x).exp();
    sign * y
}

fn gelu(x: f32) -> f32 {
    0.5 * x * (1.0 + erf(x / std::f32::consts::SQRT_2))
}

impl Model {
    pub fn load(ckpt: &str, cfg: Config) -> Result<Model, Box<dyn std::error::Error>> {
        let bytes = std::fs::read(ckpt)?;
        let st = SafeTensors::deserialize(&bytes)?;
        let mut w = HashMap::new();
        for name in st.names() {
            let t = st.tensor(name)?;
            let data = t.data();
            let mut v = Vec::with_capacity(data.len() / 4);
            for chunk in data.chunks_exact(4) {
                v.push(f32::from_le_bytes([chunk[0], chunk[1], chunk[2], chunk[3]]));
            }
            w.insert(name.to_string(), v);
        }
        let head_dim = cfg.dim / cfg.n_heads;
        Ok(Model { w, cfg, head_dim })
    }

    fn g(&self, k: &str) -> &[f32] {
        self.w.get(k).unwrap_or_else(|| panic!("falta peso: {}", k))
    }

    /// y = x @ W^T + b   con x:(t,in), W:(out,in) row-major -> y:(t,out)
    fn linear(&self, x: &[f32], t: usize, in_d: usize, w: &[f32], b: Option<&[f32]>, out_d: usize) -> Vec<f32> {
        let mut y = vec![0.0f32; t * out_d];
        for i in 0..t {
            let xr = &x[i * in_d..i * in_d + in_d];
            for o in 0..out_d {
                let wr = &w[o * in_d..o * in_d + in_d];
                let mut acc = 0.0f32;
                for k in 0..in_d {
                    acc += xr[k] * wr[k];
                }
                if let Some(bb) = b {
                    acc += bb[o];
                }
                y[i * out_d + o] = acc;
            }
        }
        y
    }

    fn layernorm(&self, x: &[f32], t: usize, dim: usize, w: &[f32], b: &[f32]) -> Vec<f32> {
        let eps = 1e-5f32;
        let mut y = vec![0.0f32; t * dim];
        for i in 0..t {
            let row = &x[i * dim..i * dim + dim];
            let mean: f32 = row.iter().sum::<f32>() / dim as f32;
            let var: f32 = row.iter().map(|v| (v - mean) * (v - mean)).sum::<f32>() / dim as f32;
            let inv = 1.0 / (var + eps).sqrt();
            for d in 0..dim {
                y[i * dim + d] = (row[d] - mean) * inv * w[d] + b[d];
            }
        }
        y
    }

    fn attention(&self, x: &[f32], t: usize, li: usize) -> Vec<f32> {
        let dim = self.cfg.dim;
        let nh = self.cfg.n_heads;
        let hd = self.head_dim;
        let p = format!("blocks.{}.attn.", li);
        let q = self.linear(x, t, dim, self.g(&format!("{}query_proj.weight", p)), Some(self.g(&format!("{}query_proj.bias", p))), dim);
        let k = self.linear(x, t, dim, self.g(&format!("{}key_proj.weight", p)), Some(self.g(&format!("{}key_proj.bias", p))), dim);
        let v = self.linear(x, t, dim, self.g(&format!("{}value_proj.weight", p)), Some(self.g(&format!("{}value_proj.bias", p))), dim);

        let scale = 1.0f32 / (hd as f32).sqrt();
        let mut out = vec![0.0f32; t * dim]; // concatenación de cabezas
        for h in 0..nh {
            let off = h * hd;
            for i in 0..t {
                // scores causales sobre j=0..=i, con softmax
                let qi = &q[i * dim + off..i * dim + off + hd];
                let mut scores = vec![0.0f32; i + 1];
                let mut maxs = f32::NEG_INFINITY;
                for j in 0..=i {
                    let kj = &k[j * dim + off..j * dim + off + hd];
                    let mut acc = 0.0f32;
                    for d in 0..hd {
                        acc += qi[d] * kj[d];
                    }
                    let s = acc * scale;
                    scores[j] = s;
                    if s > maxs {
                        maxs = s;
                    }
                }
                let mut sum = 0.0f32;
                for s in scores.iter_mut() {
                    *s = (*s - maxs).exp();
                    sum += *s;
                }
                // salida ponderada
                for d in 0..hd {
                    let mut acc = 0.0f32;
                    for j in 0..=i {
                        acc += (scores[j] / sum) * v[j * dim + off + d];
                    }
                    out[i * dim + off + d] = acc;
                }
            }
        }
        // proyección de salida
        self.linear(&out, t, dim, self.g(&format!("{}out_proj.weight", p)), Some(self.g(&format!("{}out_proj.bias", p))), dim)
    }

    fn block(&self, x: &mut Vec<f32>, t: usize, li: usize) {
        let dim = self.cfg.dim;
        let p = format!("blocks.{}.", li);
        // atención + residual
        let h = self.layernorm(x, t, dim, self.g(&format!("{}ln1.weight", p)), self.g(&format!("{}ln1.bias", p)));
        let a = self.attention(&h, t, li);
        for i in 0..x.len() {
            x[i] += a[i];
        }
        // mlp + residual
        let h = self.layernorm(x, t, dim, self.g(&format!("{}ln2.weight", p)), self.g(&format!("{}ln2.bias", p)));
        let hidden = 4 * dim;
        let mut m = self.linear(&h, t, dim, self.g(&format!("{}mlp.layers.0.weight", p)), Some(self.g(&format!("{}mlp.layers.0.bias", p))), hidden);
        for v in m.iter_mut() {
            *v = gelu(*v);
        }
        let m = self.linear(&m, t, hidden, self.g(&format!("{}mlp.layers.2.weight", p)), Some(self.g(&format!("{}mlp.layers.2.bias", p))), dim);
        for i in 0..x.len() {
            x[i] += m[i];
        }
    }

    /// Forward completo. Devuelve logits de la ÚLTIMA posición: Vec<f32> de vocab_size.
    pub fn forward_last(&self, ids: &[u32]) -> Vec<f32> {
        let dim = self.cfg.dim;
        let t = ids.len();
        let tok = self.g("tok_emb.weight");
        let pos = self.g("pos_emb.weight");
        let mut x = vec![0.0f32; t * dim];
        for (i, &id) in ids.iter().enumerate() {
            let tr = &tok[id as usize * dim..id as usize * dim + dim];
            let pr = &pos[i * dim..i * dim + dim];
            for d in 0..dim {
                x[i * dim + d] = tr[d] + pr[d];
            }
        }
        for li in 0..self.cfg.n_layers {
            self.block(&mut x, t, li);
        }
        let x = self.layernorm(&x, t, dim, self.g("ln_f.weight"), self.g("ln_f.bias"));
        // logits solo de la última posición
        let last = &x[(t - 1) * dim..t * dim];
        let head = self.g("head.weight");
        let vocab = self.cfg.vocab_size;
        let mut logits = vec![0.0f32; vocab];
        for vv in 0..vocab {
            let wr = &head[vv * dim..vv * dim + dim];
            let mut acc = 0.0f32;
            for d in 0..dim {
                acc += last[d] * wr[d];
            }
            logits[vv] = acc;
        }
        logits
    }
}

/// Lee el sidecar .json (junto al .safetensors) y arma el Config + ruta del tokenizer.
pub fn load_config(ckpt: &str) -> Result<(Config, String), Box<dyn std::error::Error>> {
    let json_path = Path::new(ckpt).with_extension("json");
    let txt = std::fs::read_to_string(&json_path)?;
    let v: serde_json::Value = serde_json::from_str(&txt)?;
    let m = &v["model"];
    let cfg = Config {
        vocab_size: m["vocab_size"].as_u64().unwrap() as usize,
        block_size: m["block_size"].as_u64().unwrap() as usize,
        dim: m["dim"].as_u64().unwrap() as usize,
        n_layers: m["n_layers"].as_u64().unwrap() as usize,
        n_heads: m["n_heads"].as_u64().unwrap() as usize,
    };
    let tok = v["tokenizer"].as_str().unwrap_or("tokenizer.json").to_string();
    Ok((cfg, tok))
}

File: src/portable/engine.py
"""
Motor de inferencia en NumPy puro (sin MLX, sin PyTorch).

Carga los pesos .safetensors entrenados con MLX y reproduce exactamente el
forward de `src/model.py`. Corre en cualquier CPU: Raspberry Pi, x86 viejo,
RK3588, laptops sin GPU... la "capa para equipos viejos" del proyecto.

Uso desde código:
    from src.portable import cargar
    m = cargar("checkpoints/yachay.safetensors")   # lee también el .json y el tokenizer
    print(m.responder("¿por qué el cielo es azul?"))

Uso desde terminal:
    python -m src.portable.run --prompt "¿por qué el cielo es azul?"
"""
import json
import math
import os

import numpy as np


def _gelu(x):
    # GELU exacto (igual que nn.GELU() de MLX): 0.5*x*(1+erf(x/sqrt(2)))
    return 0.5 * x * (1.0 + _erf(x / math.sqrt(2.0)))


def _erf(x):
    # Aproximación de Abramowitz & Stegun 7.1.26 (error < 1.5e-7). numpy puro.
    sign = np.sign(x)
    x = np.abs(x)
    t = 1.0 / (1.0 + 0.3275911 * x)
    y = 1.0 - (((((1.061405429 * t - 1.453152027) * t) + 1.421413741) * t
                - 0.284496736) * t + 0.254829592) * t * np.exp(-x * x)
    return sign * y


def _layernorm(x, w, b, eps=1e-5):
    mu = x.mean(-1, keepdims=True)
    var = x.var(-1, keepdims=True)
    return (x - mu) / np.sqrt(var + eps) * w + b


def _linear(x, w, b=None):
    # Pesos MLX/PyTorch: w shape (out, in) -> y = x @ w.T + b
    y = x @ w.T
    return y + b if b is not None else y


def _softmax(x, axis=-1):
    x = x - x.max(axis=axis, keepdims=True)
    e = np.exp(x)
    return e / e.sum(axis=axis, keepdims=True)


class GPTNumpy:
    def __init__(self, weights, cfg, tokenizer):
        self.W = weights
        self.cfg = cfg
        self.tok = tokenizer
        self.n_layers = cfg["n_layers"]
        self.n_heads = cfg["n_heads"]
        self.dim = cfg["dim"]
        self.block_size = cfg["block_size"]
        self.head_dim = self.dim // self.n_heads

    def _attn(self, x, li):
        B, T, C = x.shape
        p = f"blocks.{li}.attn."
        q = _linear(x, self.W[p + "query_proj.weight"], self.W[p + "query_proj.bias"])
        k = _linear(x, self.W[p + "key_proj.weight"], self.W[p + "key_proj.bias"])
        v = _linear(x, self.W[p + "value_proj.weight"], self.W[p + "value_proj.bias"])
        # (B,T,H,hd) -> (B,H,T,hd)
        q = q.reshape(B, T, self.n_heads, self.head_dim).transpose(0, 2, 1, 3)
        k = k.reshape(B, T, self.n_heads, self.head_dim).transpose(0, 2, 1, 3)
        v = v.reshape(B, T, self.n_heads, self.head_dim).transpose(0, 2, 1, 3)
        scale = 1.0 / math.sqrt(self.head_dim)
        scores = (q * scale) @ k.transpose(0, 1, 3, 2)         # (B,H,T,T)
        mask = np.triu(np.full((T, T), -np.inf, dtype=scores.dtype), k=1)
        scores = scores + mask
        att = _softmax(scores, axis=-1)
        out = att @ v                                          # (B,H,T,hd)
        out = out.transpose(0, 2, 1, 3).reshape(B, T, C)
        return _linear(out, self.W[p + "out_proj.weight"], self.W[p + "out_proj.bias"])

    def _block(self, x, li):
        p = f"blocks.{li}."
        h = _layernorm(x, self.W[p + "ln1.weight"], self.W[p + "ln1.bias"])
        x = x + self._attn(h, li)
        h = _layernorm(x, self.W[p + "ln2.weight"], self.W[p + "ln2.bias"])
        h = _linear(h, self.W[p + "mlp.layers.0.weight"], self.W[p + "mlp.layers.0.bias"])
        h = _gelu(h)
        h = _linear(h, self.W[p + "mlp.layers.2.weight"], self.W[p + "mlp.layers.2.bias"])
        return x + h

    def forward(self, idx):
        """idx: (B, T) enteros -> logits (B, T, vocab)."""
        B, T = idx.shape
        pos = np.arange(T)
        x = self.W["tok_emb.weight"][idx] + self.W["pos_emb.weight"][pos]
        for li in range(self.n_layers):
            x = self._block(x, li)
        x = _layernorm(x, self.W["ln_f.weight"], self.W["ln_f.bias"])
        return _linear(x, self.W["head.weight"])

    def generate(self, ids, max_new_tokens=120, temperature=0.8, top_k=40, eos_id=None, rng=None):
        rng = rng or np.random.default_rng()
        ids = list(ids)
        for _ in range(max_new_tokens):
            cond = np.array([ids[-self.block_size:]], dtype=np.int64)
            logits = self.forward(cond)[0, -1, :]
            logits = logits / max(temperature, 1e-6)
            if top_k:
                kth = np.sort(logits)[-min(top_k, logits.shape[0])]
                logits = np.where(logits < kth, -np.inf, logits)
            probs = _softmax(logits)
            nxt = int(rng.choice(len(probs), p=probs))
            ids.append(nxt)
            if eos_id is not None and nxt == eos_id:
                break
        return ids

    def responder(self, prompt, max_new_tokens=120, temperature=0.8, top_k=40,
                  seed=None, collapse_digits=False):
        rng = np.random.default_rng(seed)
        eos_id = None
        try:
            eos_id = self.tok.tk.token_to_id("<eos>")
        except Exception:
            pass
        start = self.tok.encode(f"<bos>{prompt}\n")
        out = self.generate(start, max_new_tokens, temperature, top_k, eos_id, rng)
        texto = self.tok.decode(out)
        if collapse_digits:
            # Une dígitos que el tokenizer de matemática separa ('1 1 8' -> '118')
            # y aprieta espacios dobles (cosmético).
            import re
            texto = re.sub(r"(?<=\d)\s+(?=\d)", "", texto)
            texto = re.sub(r"[ \t]{2,}", " ", texto)
        return texto


def cargar(ckpt="checkpoints/yachay.safetensors", cfg_path=None, tokenizer_path=None):
    """Carga pesos + config + tokenizer y devuelve un GPTNumpy listo para usar."""
    from safetensors.numpy import load_file
    from src.tokenizer import TokenizerBPE

    weights = {k: v.astype(np.float32) for k, v in load_file(ckpt).items()}

    if cfg_path is None:
        cfg_path = os.path.splitext(ckpt)[0] + ".json"
    with open(cfg_path, encoding="utf-8") as f:
        meta = json.load(f)
    cfg = meta["model"]

    if tokenizer_path is None:
        tokenizer_path = meta.get("tokenizer", "tokenizer.json")
    tok = TokenizerBPE.cargar(tokenizer_path)
    return GPTNumpy(weights, cfg, tok)
